### PR TITLE
implement restart policy for VM

### PR DESCRIPTION
**Pull request by cicdteam**
Restarting VM runner pods.

- by default `.spec.restartPolicy` is `Always` now
- if the runner pod is gone (manually removed or evicted by k8s control plane) the VM will go into `VmFailed` state and then runner will start again
- in case `.spec.restartPolicy: OnFailure` runner pod **will not** restart if it completed successfully

### CODE REVIEW

**Review by sharnoff — COMMENTED** [review]
There's maybe a couple nits, but given the controller rewrite I think it's fine.

The one thing I'd note is that IMO we should keep the RestartPolicy default as "Never", and instead update the control plane to explicitly set it to "Always". AFAICT it doesn't matter if the control plane sets it to Always before this is merged, because currently RestartPolicy does nothing, right?

Thoughts?

(I'll open a control plane PR to do that anyways, because it's worth being explicit)

**EDIT:** Cloud PR: https://github.com/neondatabase/cloud/pull/5146

**Comment by sharnoff** [comment]
Oh actually, turns out cloud explicitly sets it to Never - so it needs updating anyways.

**Comment by cicdteam** [comment]
> it doesn't matter if the control plane sets it to Always before this is merged, because currently RestartPolicy does nothing, right?

No, this is wrong. Currently, RestartPolicy is passed to the Pod spec and doesn't work correctly (if the pod is down, it restarts by k8s and goes to Error state at the end (because k8s just restarts, not recreates the pod))

**Comment by sharnoff** [comment]
> Currently, RestartPolicy is passed to the Pod spec

Ah right, forgot. Good catch. Added a note to the control plane PR that it must not be merged until after the VM controller update is deployed.

**Review by sharnoff — APPROVED** [review]
Hm, actually - I just read through the pod spec, turns out "Always" is the default RestartPolicy there as well. Happy to change the default to Always.

File: neonvm/controllers/virtualmachine_controller.go
Comment (tychoish) [inline neonvm/controllers/virtualmachine_controller.go:513]: my instinct is that we should avoid logging errors and then returning them: if you do this every time you get an error, you end up with one error being logged potentially many times. I don't, from this diff, how much logging we end up doing, but it seems like it could be a lot.

File: neonvm/controllers/virtualmachine_controller.go
Comment (tychoish) [inline neonvm/controllers/virtualmachine_controller.go:517]: why is this line split? we have a bunch of quite long lines, and I think if we're going to have long lines, we should avoid splits like this except when it improves clarity, and I'm not sure that it makes sense in this case.

File: neonvm/controllers/virtualmachine_controller.go
Comment (tychoish) [inline neonvm/controllers/virtualmachine_controller.go:520]: are TODOs tracked somewhere (issues, centrally?) on what time scale are we expecting to resolve this? 

also a comment about why 1 second was chosen and what would happen (assuming the refactor happens after we merge this) if we changed this in either direction.

File: neonvm/controllers/virtualmachine_controller.go
Comment (tychoish) [inline neonvm/controllers/virtualmachine_controller.go:539]: I think/hope that `apierrors.IsNotFound` will return false if `err == nil` so it makes this easier to read if you delete the `err != nil` on this line.

File: neonvm/controllers/virtualmachine_controller.go
Comment (tychoish) [inline neonvm/controllers/virtualmachine_controller.go:532]: we have this code in two places. can we put it into a method somewhere?

File: neonvm/controllers/virtualmachine_controller.go
Comment (tychoish) [inline neonvm/controllers/virtualmachine_controller.go:822]: I don't quite understand this one.

**Review by tychoish — COMMENTED** [review]
Nothing here that should block merging, just style and clarity questions.

File: neonvm/controllers/virtualmachine_controller.go
Comment (cicdteam) [reply neonvm/controllers/virtualmachine_controller.go:515]: Recorder Events have the same style over whole code. It was initially generated by kubebuilder and i think it's good.

File: neonvm/controllers/virtualmachine_controller.go
Comment (cicdteam) [reply neonvm/controllers/virtualmachine_controller.go]: `time.Sleep` has been removed as it now works without delay. In my tests with the wrong vm phase cleanup, this resulted in duplicate runner pods, but now everything is fine.

File: neonvm/controllers/virtualmachine_controller.go
Comment (cicdteam) [reply neonvm/controllers/virtualmachine_controller.go]: I originally decided not to add another method to the VirtualMachine type because cleanup only happens in those two places, but now I've added Cleanup().

File: neonvm/controllers/virtualmachine_controller.go
Comment (cicdteam) [reply neonvm/controllers/virtualmachine_controller.go]: deleted

**Pull request merged**
